rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -496,4 +567,134 @@ void Recorder::warn_if_new_qos_for_subscribed_topic(const std::string & topic_na
   }
 }
 
+///////////////////////////////
+// Recorder public interface
+
+Recorder::Recorder(
+  const std::string & node_name,
+  const rclcpp::NodeOptions & node_options)
+: rclcpp::Node(node_name, node_options)
+{
+  // TODO(karsten1987): Use this constructor later with parameter parsing.
+  // The reader, storage_options as well as record_options can be loaded via parameter.
+  // That way, the recorder can be used as a simple component in a component manager.
+  throw rclcpp::exceptions::UnimplementedError();
+}
+
+Recorder::Recorder(
+  std::shared_ptr<rosbag2_cpp::Writer> writer,
+  const rosbag2_storage::StorageOptions & storage_options,
+  const rosbag2_transport::RecordOptions & record_options,
+  const std::string & node_name,
+  const rclcpp::NodeOptions & node_options)
+: Recorder(
+    std::move(writer),
+#ifndef _WIN32
+    std::make_shared<KeyboardHandler>(false),
+#else
+    // We don't have signal handler option in constructor for windows version
+    std::shared_ptr<KeyboardHandler>(new KeyboardHandler()),
+#endif
+    storage_options,
+    record_options,
+    node_name,
+    node_options)
+{}
+
+Recorder::Recorder(
+  std::shared_ptr<rosbag2_cpp::Writer> writer,
+  std::shared_ptr<KeyboardHandler> keyboard_handler,
+  const rosbag2_storage::StorageOptions & storage_options,
+  const rosbag2_transport::RecordOptions & record_options,
+  const std::string & node_name,
+  const rclcpp::NodeOptions & node_options)
+: rclcpp::Node(node_name, rclcpp::NodeOptions(node_options)
+    .start_parameter_event_publisher(false)
+    .parameter_overrides({rclcpp::Parameter("use_sim_time", record_options.use_sim_time)})),
+  pimpl_(std::make_unique<RecorderImpl>(
+      this, std::move(writer), keyboard_handler,
+      storage_options, record_options))
+{}
+
+Recorder::~Recorder()
+{}
+
+void
+Recorder::record()
+{
+  pimpl_->record();
+}
+
+const std::unordered_set<std::string> &
+Recorder::topics_using_fallback_qos() const
+{
+  return pimpl_->topics_warned_about_incompatibility_;
+}
+
+const std::unordered_map<std::string, std::shared_ptr<rclcpp::GenericSubscription>> &
+Recorder::subscriptions() const
+{
+  return pimpl_->subscriptions_;
+}
+
+const rosbag2_cpp::Writer &
+Recorder::get_writer_handle()
+{
+  return *pimpl_->writer_;
+}
+
+void
+Recorder::pause()
+{
+  pimpl_->pause();
+}
+
+void
+Recorder::resume()
+{
+  pimpl_->resume();
+}
+
+void
+Recorder::toggle_paused()
+{
+  pimpl_->toggle_paused();
+}
+
+bool
+Recorder::is_paused()
+{
+  return pimpl_->is_paused();
+}
+
+std::unordered_map<std::string, std::string>
+Recorder::get_requested_or_available_topics()
+{
+  return pimpl_->get_requested_or_available_topics();
+}
+
+rosbag2_cpp::Writer &
+Recorder::get_writer()
+{
+  return *pimpl_->writer_;
+}
```

Comment:
We already have 
```cpp
const rosbag2_cpp::Writer &
Recorder::get_writer_handle()
{
  return *pimpl_->writer_;
}
```
Need to keep one of them other wise it will be kind of confusion in API

Reply:
This is a non-const `protected` version so that subclasses can use the writer directly - I'm just trying not to reduce the ability level of any existing subclass use cases. `get_writer_handle` is  exposed publicly. I think it's only used in tests, though (in the rosbag2 codebase). Those tests don't actually need to do it this way - should I rewrite those and remove `get_writer_handle()`?

Reply:
> should I rewrite those and remove get_writer_handle()?

Yes please.

Reply:
After some consideration if it will  require to add derived  recorder class with exposed `Recorder::get_writer()` API - let's keep it as is. I am actually missed a point that one of them is in protected space.